### PR TITLE
Math/vector implementation

### PR DESCRIPTION
**Pull request by DRossen**
Cleaned up implementation of Vector3 and Vector3 float/int versions.
Also added some tests.

### CODE REVIEW

File: Source/CoreTests/Vector3Tests.cpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    Vector3 v2 = Vector3::Zero;
+    EXPECT_FLOAT_EQ(v2.x, 0.0f);
+    EXPECT_FLOAT_EQ(v2.y, 0.0f);
+    EXPECT_FLOAT_EQ(v2.z, 0.0f);
+
+    float arr[3] = { 7.0f, 8.0f, 9.0f };
+    Vector3 v3(arr);
+    EXPECT_FLOAT_EQ(v3.x, 7.0f);
+    EXPECT_FLOAT_EQ(v3.y, 8.0f);
+    EXPECT_FLOAT_EQ(v3.z, 9.0f);
```
Comment (OlleKReutercrona) [inline Source/CoreTests/Vector3Tests.cpp]: ~What's the difference between v1 and v3 in this case?~ Array construction 👍 
A good habit to have in these cases is to re-use the values from the array as a expected result, especially when dealing with strings.
EXPECT_FLOAT_EQ(v3.x, arr[0]);

File: Source/CoreTests/Vector2Tests.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+TEST(Vector2Tests, LengthAndLengthSquared) {
+    Vector2 v(3.0f, 4.0f);
+    EXPECT_FLOAT_EQ(v.LengthSquared(), 25.0f);
+    EXPECT_FLOAT_EQ(v.Length(), 5.0f);
+}
+
+TEST(Vector2Tests, DotProduct) {
+    Vector2 a(1.0f, 2.0f);
+    Vector2 b(3.0f, 4.0f);
+    EXPECT_FLOAT_EQ(a.Dot(b), 11.0f); // 1*3 + 2*4
```
Comment (OlleKReutercrona) [inline Source/CoreTests/Vector2Tests.cpp]: Nice that's you're explaining the math in the expected result 👍

File: Source/CoreTests/Vector2Tests.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+TEST(Vector2Tests, Normalize) {
+    Vector2 v(3.0f, 4.0f);
+    Vector2 n = v.Normalized();
+    EXPECT_NEAR(n.Length(), 1.0f, 1e-6f);
+    EXPECT_NEAR(n.x, 0.6f, 1e-6f);
+    EXPECT_NEAR(n.y, 0.8f, 1e-6f);
```
Comment (OlleKReutercrona) [inline Source/CoreTests/Vector2Tests.cpp]: `1e-6f` can be a global constexpr to be shared in the relevant places

File: Source/CoreTests/Vector2Tests.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    Vector2 a(2, 3);
+    Vector2 b(1, -1);
+
+    EXPECT_EQ(a + b, Vector2(3, 2));
+    EXPECT_EQ(a - b, Vector2(1, 4));
+
+    EXPECT_EQ(a * 2.0f, Vector2(4, 6));
+    EXPECT_EQ(2.0f * a, Vector2(4, 6));
+    EXPECT_EQ(a / 2.0f, Vector2(1, 1.5f));
+}
```
Comment (OlleKReutercrona) [inline Source/CoreTests/Vector2Tests.cpp]: Nice but is missing the assignment operators too?
+=
-=
/=
*=

File: Source/CoreTests/Vector2Tests.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
+#include <gtest/gtest.h>
+#include "Math/Vector2.h"
+
+TEST(Vector2Tests, ConstructorFromArray) {
+    float arr[2] = { 3.5f, -2.0f };
+    Vector2 v(arr);
+    EXPECT_FLOAT_EQ(v.x, 3.5f);
+    EXPECT_FLOAT_EQ(v.y, -2.0f);
+}
```
Comment (OlleKReutercrona) [inline Source/CoreTests/Vector2Tests.cpp]: Nice but you should also test all the other constructors too

File: Source/CoreTests/Vector2Tests.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+TEST(Vector2iTests, OperatorsArithmetic) {
+    Vector2i a(2, 3);
+    Vector2i b(1, -1);
+
+    EXPECT_EQ(a + b, Vector2i(3, 2));
+    EXPECT_EQ(a - b, Vector2i(1, 4));
+
+    EXPECT_EQ(a * 2, Vector2i(4, 6));
+    EXPECT_EQ(2 * a, Vector2i(4, 6));
+    EXPECT_EQ(a / 2, Vector2i(1, 1));
```
Comment (OlleKReutercrona) [inline Source/CoreTests/Vector2Tests.cpp]: Is also missing assignment operators

File: Source/CoreTests/Vector3Tests.cpp [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    Vector3 scaled = v * 2.0f;
+    EXPECT_FLOAT_EQ(scaled.x, 4);
+    EXPECT_FLOAT_EQ(scaled.y, -8);
+    EXPECT_FLOAT_EQ(scaled.z, 12);
+
+    Vector3 divided = scaled / 2.0f;
+    EXPECT_FLOAT_EQ(divided.x, 2);
+    EXPECT_FLOAT_EQ(divided.y, -4);
+    EXPECT_FLOAT_EQ(divided.z, 6);
+}
```
Comment (OlleKReutercrona) [inline Source/CoreTests/Vector3Tests.cpp]: You should also make tests to make sure that we catch division by zero. You can check how I implemented "death-tests" in matrix3x3

File: Source/CoreTests/Vector3Tests.cpp (hunk 1 above)
Comment (OlleKReutercrona) [reply Source/CoreTests/Vector3Tests.cpp]: Also, add the other constructors to this test

File: Source/CoreTests/Vector3Tests.cpp [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    EXPECT_EQ(v2.x, 0);
+    EXPECT_EQ(v2.y, 0);
+    EXPECT_EQ(v2.z, 0);
+
+    int arr[3] = { 7, 8, 9 };
+    Vector3i v(arr);
+    EXPECT_EQ(v.x, 7);
+    EXPECT_EQ(v.y, 8);
+    EXPECT_EQ(v.z, 9);
+}
```
Comment (OlleKReutercrona) [inline Source/CoreTests/Vector3Tests.cpp]: Is missing some constructors

File: Source/Core/Math/Vector2.h [hunk 9]
```diff
@@ -1,119 +1,248 @@
...
+        return a; 
+    }
+    friend Vector2 operator*(Vector2 a, float s) noexcept { 
+        a *= s;
+        return a; 
+    }
+    friend Vector2 operator*(float s, Vector2 a) noexcept {
+        a *= s; 
+        return a; 
+    }
```
Comment (OlleKReutercrona) [inline Source/Core/Math/Vector2.h:94]: Is there a reason for these being `friend` declared? 🤔

File: Source/Core/Math/Vector2.h [hunk 10]
```diff
@@ -1,119 +1,248 @@
...
+    float Dot(const Vector2 &other) const noexcept { 
+        return x * other.x + y * other.y; 
+    }
+    [[nodiscard]] Vector2 Normalized() const noexcept {
+        float lsq = LengthSquared();
+        if (lsq <= 1e-8f) { return Vector2::Zero; }
+        return (*this) * (1.0f / sqrtf(lsq));
+    }
+    [[nodiscard]] Vector2 Reflect(const Vector2 &normal) const noexcept {
+        assert(fabs(normal.LengthSquared() - 1.0f) < 1e-3f && "Vector3::Reflect received non-normalized normal");
```
Comment (OlleKReutercrona) [inline Source/Core/Math/Vector2.h:122]: Nice! Don't forget to create a test for this too

File: Source/Core/Math/Vector2.h [hunk 11]
```diff
@@ -1,119 +1,248 @@
...
+        y -= rhs.y;
+        return *this;
+    }
+    Vector2i &operator*=(int s) noexcept {
+        x *= s;
+        y *= s;
+        return *this;
+    }
+    Vector2i &operator/=(int s) noexcept {
+        assert(s != 0 && "Division by zero in Vector2i::operator/=");
```
Comment (OlleKReutercrona) [inline Source/Core/Math/Vector2.h:199]: Nice! Don't forget to test this

**Review by OlleKReutercrona — APPROVED** [review]
Really nice, glad to see you start writing test! Added some minor comments here and there but I'd like you to implement the missing tests I suggested before you merge 👍 
Good job!

File: Source/Core/Math/Vector2.h (hunk 9 above)
Comment (DRossen) [reply Source/Core/Math/Vector2.h:94]: friend is to allow them to be defined within the class, to keep them along the other operators, it has no other purpose here

**Comment by DRossen** [comment]
image

File: Source/CoreTests/premake5.lua [hunk 12]
```diff
@@ -59,10 +59,15 @@ project(NAME)
         flags {
 			"MultiProcessorCompile"
         }
+		
+		disablewarnings {
+            "4723" -- division by zero
+        }
```
Comment (OlleKReutercrona) [inline Source/CoreTests/premake5.lua:65]: Really nice!
